src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+      #SERVICE ".grpc_client." #METHOD,                                       \
+      request,                                                                \
+      callback,                                                               \
+      method_timeout_ms))
+
+// Define a void retryable RPC client method.
+#define VOID_RETRYABLE_RPC_CLIENT_METHOD(                                        \
+    retryable_rpc_client, SERVICE, METHOD, rpc_client, method_timeout_ms, SPECS) \
+  void METHOD(const METHOD##Request &request,                                    \
+              const ClientCallback<METHOD##Reply> &callback) SPECS {             \
+    INVOKE_RETRYABLE_RPC_CALL(retryable_rpc_client,                              \
+                              SERVICE,                                           \
+                              METHOD,                                            \
+                              request,                                           \
+                              callback,                                          \
+                              rpc_client,                                        \
+                              method_timeout_ms);                                \
+  }
+
+/// \class Executor
+/// Executor saves operation and support retries.
+class Executor {
+ public:
+  Executor(std::function<void(const ray::Status &)> abort_callback)
+      : abort_callback_(std::move(abort_callback)) {}
+
+  /// This function is used to execute the given operation.
+  ///
+  /// \param operation The operation to be executed.
+  void Execute(std::function<void()> operation) {
+    operation_ = std::move(operation);
+    operation_();
+  }
+
+  /// This function is used to retry the given operation.
+  void Retry() { operation_(); }
+
+  void Abort(const ray::Status &status) { abort_callback_(status); }
+
+ private:
+  std::function<void(ray::Status)> abort_callback_;
+  std::function<void()> operation_;
+};
+
+/**
+ * The client makes RPC calls through the provided underlying grpc client.
+ * If the call goes through, the user provided callback is invoked.
+ * If the call fails due to transient network error, it is added to a retry queue.
+ * The client waits for the grpc channel reconnection to resend the requests.
+ * If the total number of request bytes in the queue exceeds max_pending_requests_bytes,
+ * the thread is blocked until some requests are resent.
+ * If a call's timeout_ms reaches during retry, its callback is called with
+ * Status::TimedOut. If the whole client does not reconnect within
+ * server_unavailable_timeout_seconds, server_unavailable_timeout_callback is invoked.
+ * When all callers of the client release the shared_ptr of the client, the client
+ * destructor is called and the client is shut down.
+ */
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
```

Comment:
I'm not sure why do we place the whole implementation in the header file?
it's not a templated class, right?